asapdiscovery-data/asapdiscovery/data/state_expanders/protomer_expander.py

```diff
@@ -0,0 +1,208 @@
+import os
+import subprocess
+import tempfile
+from typing import Literal
+
+from asapdiscovery.data.openeye import (
+    load_openeye_sdfs,
+    oechem,
+    oequacpac,
+    save_openeye_sdfs,
+)
+from asapdiscovery.data.schema_v2.ligand import Ligand
+from asapdiscovery.data.state_expanders.state_expander import StateExpanderBase
+from pydantic import Field
+
+
+class ProtomerExpander(StateExpanderBase):
+    """
+    Expand a molecule to protomers using OpenEye reasonable protomer state enumeration.
+
+    Note:
+        The input molecule is included in the output.
+    """
+
+    expander_type: Literal["ProtomerExpander"] = "ProtomerExpander"
+
+    def _provenance(self) -> dict[str, str]:
+        return {
+            "oechem": oechem.OEChemGetVersion(),
+            "quacpac": oequacpac.OEQuacPacGetVersion(),
+        }
+
+    def _expand(self, ligands: list[Ligand]) -> list[Ligand]:
+        expanded_states = []
+        provenance = self.provenance()
+        for parent_ligand in ligands:
+            oemol = parent_ligand.to_oemol()
+            for protomer in oequacpac.OEGetReasonableProtomers(oemol):
+                fmol = oechem.OEMol(protomer)
+                # copy the ligand properties over to the new molecule, we may want to have more fine grained control over this
+                # down the track.
+                protomer_ligand = Ligand.from_oemol(fmol, **parent_ligand.dict())
+                if protomer_ligand.fixed_inchikey != parent_ligand.fixed_inchikey:
+                    # only add tags to new microstates of the input molecule
```

Comment:
Ah this is much nicer.

Reply:
Thanks it took a while to get there but I think its worth it.